onnxruntime/core/providers/coreml/builders/impl/cast_op_builder.cc

```diff
@@ -18,14 +19,49 @@ class CastOpBuilder : public BaseOpBuilder {
 
   bool HasSupportedInputsImpl(const Node& node, const OpBuilderInputParams& input_params,
                               const logging::Logger& logger) const override;
+
+ public:
+  bool SupportsMLProgram() const override { return true; }
 };
 
-Status CastOpBuilder::AddToModelBuilderImpl(ModelBuilder& /* model_builder */,
-                                            const Node& /* node */,
-                                            const logging::Logger& /* logger */) const {
-  // This is a special handling case for ArgMax Op, where argmax is followed by a cast to int32 type.
-  // The ArgMax is fused with the Cast node and produces an int32 output.
-  // Cast node is not provided in CoreML model, so we're skipping adding the Cast node here.
+Status CastOpBuilder::AddToModelBuilderImpl([[maybe_unused]] ModelBuilder& model_builder,
+                                            [[maybe_unused]] const Node& node,
+                                            [[maybe_unused]] const logging::Logger& logger) const {
+// This is a special handling case for ArgMax Op, where argmax is followed by a cast to int32 type.
+// The ArgMax is fused with the Cast node and produces an int32 output.
+#if defined(COREML_ENABLE_MLPROGRAM)
+  if (model_builder.CreateMLProgram()) {
+    using namespace CoreML::Specification::MILSpec;
+    // https://apple.github.io/coremltools/source/coremltools.converters.mil.mil.ops.defs.html#coremltools.converters.mil.mil.ops.defs.iOS15.elementwise_unary.cast
+
+    std::unique_ptr<Operation> op = model_builder.CreateOperation(node, "cast");
+    AddOperationInput(*op, "x", node.InputDefs()[0]->Name());
+    NodeAttrHelper helper(node);
+    auto cast_to_type = helper.Get("to", ONNX_NAMESPACE::TensorProto::UNDEFINED);
+    std::string to_dtype = "";
+    if (cast_to_type == ONNX_NAMESPACE::TensorProto::INT32 || cast_to_type == ONNX_NAMESPACE::TensorProto::INT64) {
+      to_dtype = "int32";
+      // CoreML doesn't support int64 while onnx used int64 in index input/output
+      // we have two solutions for this: 1) if cast is the first/last node in graph partition,
+      // we will convert the input/output to int32 forcely
+      // 2) if cast is a middle node of graph-parition, we just treat int64 as int32.
+      cast_to_type = ONNX_NAMESPACE::TensorProto::INT32;
+    } else if (cast_to_type == ONNX_NAMESPACE::TensorProto::FLOAT) {
+      to_dtype = "fp32";
+    } else if (cast_to_type == ONNX_NAMESPACE::TensorProto::FLOAT16) {
+      to_dtype = "fp16";
+    } else if (cast_to_type == ONNX_NAMESPACE::TensorProto::BOOL) {
+      to_dtype = "bool";
+    } else {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Unsupported cast type: ", cast_to_type);
+    }
+
+    AddOperationInput(*op, "dtype", model_builder.AddScalarConstant(op->type(), "dtype", std::string(to_dtype)));
+    AddOperationOutput(*op, *node.OutputDefs()[0], cast_to_type);
+    model_builder.AddOperation(std::move(op));
```

Comment:
Do we need a check here to handle the no-op if input data type is TensorProto::INT64 or TensorProto::INT32 and the cast_to_type is int32 by inserting an Identity node? 

CoreML might detect that sort of no-op Cast and remove that itself though. Not sure if it would do that during model save and we could check the generated model to make sure that was the case. If it is we could just add a comment saying we don't need to handle this case.

Reply:
Sure, we can add it explicitly. I assume that CoreML will do it internaly